### DISCUSSION TITLE
Issues with .thr files. Upload Gcode directly?

### DISCUSSION THREAD

**Discussion by jirwin69** [rdudhagra/Sand-Table#25] [Q&A] [answered]
First off, thank you so much for creating and sharing this build guide- its amazing.
I used it to build a 6ft long computer desk and im so close to being finished but ran into a small issue. When I create tracks in sandity with my rectangular machine dimensions and export as .thr, importing the same file straight back into sandify will show that the path is way smaller than the original preview. Uploading it to the table with also plots this smaller image, along with rotating the entire track 90 degrees.

If i export the track as gcode and then import, sandify keeps everything just as it was originally.

My question here is:
1. Have you come across this before? Seems like if Sandify does this to every .thr export there's no way I'm the only one affected by it.
2. What would I need to change in the website in order to upload the gcode file directly instead of relying on it converting a .thr to gcode?

![20211206_205036](https://user-images.githubusercontent.com/95666505/144959156-f518fc14-bc84-4adb-9841-67ecc06f135c.jpg)
![20211206_205058](https://user-images.githubusercontent.com/95666505/144959161-a22bab17-44f1-4bf0-a52c-77c725804c8a.jpg)
![image2](https://user-images.githubusercontent.com/95666505/144959573-bb1b497a-4f7f-44cc-bb7f-d97544108782.png)
![image-1](https://user-images.githubusercontent.com/95666505/144959596-2f0d1635-8b0c-41c4-8653-b0355ea76f59.png)
![Screenshot (231)](https://user-images.githubusercontent.com/95666505/144961101-4fdf319d-8670-4bb4-8a69-145bffcbbdad.png)

**Comment by rdudhagra** [comment, marked as answer]
Hey! 

First off, nice table by the way!

`.thr` files are scaled down to a range from 0-1 as far as I remember (this would explain why importing it back into sandify results in a different scale—sandify just chooses some default I guess). That being said, I haven't used sandify much so not sure if they do things differently. As for the scaling issue, it looks like the dimensions of the table in the website config might be off. Have you modified the table dimensions in `Website/.env`? 

Also, what happens when you actually try playing the file from the website? Does the table draw the pattern 90 degrees off? If that's the case, you might need to flip your axes, which you could either do in hardware by swapping some wires or by modifying the `printRectangularCommand` function in `Website/server/routes/process_theta_rho.js` (swap `X` and `Y` in the calls to `write` to rotate). If the pattern is flipped, you can replace the value being outputted to the gcode with `X_SIZE - fitCoord[0].toFixed(3)` (and similarly for `Y`).

### If you just want to upload gcode, modify `Website/server/routes/process_upload.js` as follows:
- In the following code snippet:
```javascript
if (!thrFile.name.endsWith(".thr")) {
    return res.status(400).send("Invalid file type");
}
```
Modify it to also allow `.gcode` files.
- Then, you want to create an `if` statement to case on whether you're processing `gcode` or `thr` files. If you just have a `gcode`, move the file using a call similar to that on line 42 of that file (`thrFile.mv(__dirname + "/../../files/" + thrFile.name, function (err) { ...`). Make sure you call `res.send('File uploaded!');` somewhere inside that callback function. Else (we're processing a `.thr` file), run the existing code as usual.
- Last, modify `Website/src/app/components/Tracks.jsx` to allow uploading `.gcode` files, so replace 
```javascript
<input
  type="file"
  accept=".thr"
  onChange={(e) => this.props.uploadNew(e)}
  hidden
/>
```
with
```javascript
<input
  type="file"
  accept=".thr,.gcode"
  onChange={(e) => this.props.uploadNew(e)}
  hidden
/>
```
You'll need to run `npm run build` from the terminal in the `Website` directory to rebuild the website after the change.

As a note which I'm sure you are aware of, if you upload your own gcode, you could potentially damage your table by commanding it something that's outside its bounds/etc. `.thr` files as I mentioned above have a fixed scale (radius from 0-1, theta is in radians and can be anything), and each table is responsible for scaling it up/down as needed.

Hopefully that helps a bit...let me know if this doesn't make sense/if you have any more questions. Good luck!

**Comment by iwannabswiss** [comment]
> When I create tracks in sandity with my rectangular machine dimensions and export as .thr, importing the same file straight back into sandify will show that the path is way smaller than the original preview. Uploading it to the table with also plots this smaller image, along with rotating the entire track 90 degrees.
> 
> If i export the track as gcode and then import, sandify keeps everything just as it was originally.
> 
> My question here is:
> 
> 1. Have you come across this before? Seems like if Sandify does this to every .thr export there's no way I'm the only one affected by it.
> 2. What would I need to change in the website in order to upload the gcode file directly instead of relying on it converting a .thr to gcode?

You are not alone.  I recently put the first round of finishing touches on my table and have the same issue.  I reached out to the developer, and we determined that it is an issue with Sandify.  Not sure how long until it'll be addressed, but I've already forked this build and implemented the change for gcode, plus a few other tidbits.

**Reply by gabhalling** [reply]
Hi,

I just swapped some of the default build files for the ones you modified in your fork, but now I only manage to upload .gcode files and .thr files make the website crash... Think you could give me a hand?

![image](https://github.com/user-attachments/assets/65153349-9558-4333-91f0-8e767d8a8fb5)

Thank you,

Gabriel

**Reply by rdudhagra** [reply]
I think you need to modify line 205 of `process_theta_rho.js` in @iwannabswiss's fork as such:

> [Line 205]
```js
    var inStream = fs.createReadStream(__dirname + "/../../files/" + filename + ".thr")
```
to
```js
    var inStream = fs.createReadStream(__dirname + "/../../files/org" + filename + ".thr")
```

**Reply by rdudhagra** [reply]
Another change in `playlist_manager.js`:
> [Line 18]
```js
    fs.mkdirSync(orgDir);
```
to
```js
    if (!fs.existsSync(orgDir))
        fs.mkdirSync(orgDir);
```
